### PR TITLE
make sure the expected delay to happen

### PR DESCRIPTION
**Pull request by QuinnWang**
make sure the delay in the release_clock_and_wait in i2c_master.xc to happen, and same idea for the delay of the wait_for_clock_high in the i2c_master_single_port.xc.

### CODE REVIEW

**Comment by xross** [comment]
I believe this is already fixed in #79

**Comment by QuinnWang** [comment]
> I believe this is already fixed in #79

Seems no, this problem is based on the #79 repeat start fix, and please check the description as below in detail:

If the clock stretching happening, for example, the scl rising edge delays for 1 clock period, the original "tmr when timerafter(fall_time + delay) :> time" will not happen actually.
If the example above does happen, the timer value "time" in the original code is actually the scl rising edge time, then suggest to update the reference time point - fall_time - to a new value, set it to one low_period_ticks ago to be as an expected last scl fall edge time, and do the delay again to make sure the delay will indeed happen.

To verify, just take the last restart fix as an example, if the scl rising edge delays for 1 clock period, the delay code in the release_clock_and_wait doesn't do the delay actually, then the following drive SDA low will run right after the scl released by I2C slave to high, we can see the restart problem again, because the previous restart bug is just the compute_bus_off_ticks delay in the release_clock_and_wait is almost equal to the compute_low_period_ticks before calling release_clock_and_wait, then no sufficient time for between scl goes high and sda goes low.
With the changes, if go through/review with the clock stretching happening cases, since the delay should happen as expected, then no problem for the clock stretching cases.

And during understanding above, I found if changing the fall_time to last_scl_fall_edge_time, the readability of the code will become friendly and logic straight forward, then just changed as well.
Then found if following the original design idea that the fall_time is last_scl_fall_edge_time, the restart fix should need to be changed slightly, so updated in the propose changes too.

**Comment by shuchitak** [comment]
I have gone through these changes and they make sense to me. Without these changes, both check_clock_high_time() and check_setup_start_time() checks fail for the clock stretching test case (test_master_clock_stretch.py) and they pass with these changes (Tested in https://github.com/xmos/lib_i2c/pull/105).

**Note that I had to reduce the expected_speed in the test_master_clock_stretch.py since doing the required delay between the slave releasing SCL and driving data on SDA means that the detected speed on I2C would be slower than before.**

File: lib_i2c/src/i2c_master.xc [hunk 1]
```diff
@@ -68,7 +69,8 @@ static void release_clock_and_wait(
   // executing then the clock needs to be adjusted
   const int wake_up_ticks = 10;
   if (time > fall_time + delay + wake_up_ticks) {
```
Comment (xross) [inline lib_i2c/src/i2c_master.xc:71]: Isn't there an issue if time wraps here? Should it not use timeafter() from xs1.h?

**Review by xross — CHANGES_REQUESTED** [review]
Agree variable name restore was the right thing to do here. 
I left a small comment about a comment. 

This PR is has 10 commits for relatively small change (including an intermediate step for where a variable is renamed). I think it should be squashed to one commit.

File: lib_i2c/src/i2c_master.xc [hunk 2]
```diff
@@ -111,13 +113,13 @@ static void inline high_pulse(
...
   tmr when timerafter(fall_time + compute_low_period_ticks(kbits_per_second)) :> void;
-  release_clock_and_wait(p_scl, fall_time, (bit_time * 3) / 4);
+  release_clock_and_wait(p_scl, fall_time, (bit_time * 3) / 4, kbits_per_second);
   fall_time = fall_time + bit_time;
   tmr when timerafter(fall_time) :> void;
   p_scl <: 0;
 }
 
-/** Output a start bit. The function returns the 'fall time' i.e. the
+/** Output a start bit. The function returns/updates the 'last SCL falling edge time' i.e. the
```
[the remark this answers is not in the document]
Comment (shuchitak) [reply lib_i2c/src/i2c_master.xc:122]: fixed

**Review by xross — APPROVED** [review]

**Pull request merged**